src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +43,189 @@ class InTotoV1Statement(TypedDict):
+        )
+
+    subjects_payload = payload.get("subject")
+    if subjects_payload is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'subject' of the in-toto statement is missing.",
+        )
+    if not isinstance(subjects_payload, list):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'subject' in the in-toto statement is invalid: expecting a list.",
+        )
+
+    for subject_json in subjects_payload:
+        validate_intoto_subject(subject_json)
+
+    predicate_type = payload.get("predicateType")
+    if predicate_type is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'predicateType' of the in-toto statement is missing.",
+        )
+
+    if not isinstance(predicate_type, str):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'predicateType' in the in-toto statement is invalid: expecting a string."
+        )
+
+    predicate = payload.get("predicate")
+    if predicate is not None and not isinstance(predicate, dict):
+        raise ValidateInTotoPayloadError(
+            "The value attribute 'predicate' in the in-toto statement is invalid: expecting an object.",
+        )
+
+    return True
+
+
+def validate_intoto_subject(subject: JsonType) -> TypeGuard[InTotoV1ResourceDescriptor]:
+    """Validate a single subject in the in-toto statement.
+
+    See specification: https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md
+
+    Parameters
+    ----------
+    subject : JsonType
+        The JSON element representing a single subject.
+
+    Returns
+    -------
+    TypeGuard[InTotoV1ResourceDescriptor]
+        ``True`` if the subject element is valid, in which case its type is narrowed to an
+        ``InTotoV1ResourceDescriptor``; ``False`` otherwise.
+
+    Raises
+    ------
+    ValidateInTotoPayloadError
+        When the payload does not follow the expecting schema.
+    """
+    if not isinstance(subject, dict):
+        raise ValidateInTotoPayloadError(
+            "A subject in the in-toto statement is invalid: expecting an object.",
+        )
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri = _validate_property(subject, "uri", lambda x: isinstance(x, str))
+    content = _validate_property(subject, "content", lambda x: isinstance(x, str))
+    digest = _validate_property(subject, "digest", is_valid_digest_set)
+    if not any([uri, content, digest]):
+        raise ValidateInTotoPayloadError(
+            "One of 'uri', 'digest', or 'content' must be present and valid within 'subject'."
+        )
+
+    _validate_property(subject, "name", lambda x: isinstance(x, str))
+    _validate_property(subject, "downloadLocation", lambda x: isinstance(x, str))
+    _validate_property(subject, "mediaType", lambda x: isinstance(x, str))
+    _validate_property(subject, "annotations", is_valid_annotation_map)
```

Comment:
Wouldn't a lambda be enough in this case, since you only need to check with `isinstance` and this is consistent with the rest of the function?

```suggestion
    _validate_property(subject, "annotations", lambda x: isinstance(x, dict))
```